### ISSUE TITLE
NetworkObjects of deleted GameObjects still being referenced in Observer Code

### ISSUE THREAD

**Issue by Shane-SDK** [FirstGearGames/FishNet#803]
**General**
Unity version: **2023.2.20f1**
Fish-Networking version: **4.5.2R**
Discord link: **[Message](https://discord.com/channels/424284635074134018/1034477094731784302/1300560688673980416)**

**Description**
NetworkObjects that are spawned at runtime are not being cleaned up properly between scene loads. 

Specifically old GameObjects which were deleted between scene loads still have their NetworkObject references in the nobCache retrieved by CollectionCaches<NetworkObject>.RetrieveList() _(ServerObjects.Observers.cs - Line 342)_ when rebuilding observers. (In 2nd screenshot)

This behavior happens both for Host/Non-host network connections. This also only occurs for NetworkObjects which were spawned at runtime using ServerManager.Spawn(...). This hasn't happened for scene NetworkObjects. I've also tried the various scene loading methods such as loading Globally, loading using the Fish SceneManager API, as well as loading/unloading the scenes with Unity's SceneManagement API manually. The behavior is present with all three methods.

In my setup I spawn a NetworkObject immediately on Scene load, at some point a NetworkConnection will load the scene, therefore loading the runtime-spawned NetworkObject. This isn't a problem for the first scene load, but if I repeat the process the old NetworkObject will still be referenced, despite no longer having an underlying GameObject, as per the prior scene unload. 

I've tested this on a few releases
**4.5.2R** - Bug exists
**4.5.1R** - Bug exists
**4.5.0R** - No bug, works as intended
**4.4.6R** - No bug, works as intended

**Replication**
1. Load a scene on the server, then spawn a Network GameObject at runtime.
2. Load a new scene on the server, ensure old scene is unloaded
3. Add a NetworkConnection to the new scene
4. Server will rebuild observers using the old NetworkObject spawned in the first step, for the NetworkConnection. During this process it will access members of it's GameObject, which no longer exists

**Expected behavior**
The nobCache shouldn't contain NetworkObjects whose GameObject was removed, as the function accesses members of the null GameObject through the NetworkObject.

**Screenshots**
**Immediate exception being raised**
![image](https://github.com/user-attachments/assets/4eac81d4-5a07-40ea-9366-f3440e13de50)

**The NetworkObjects being evaluated in observer code**
![image](https://github.com/user-attachments/assets/70e93708-4c2a-4a5b-89ce-a7f629f46f24)

**Comment by FirstGearGames** [comment]
Can you make a repo of this please. There was definitely a release where
this was fixed but might be another condition elsewhere.

**Comment by FirstGearGames** [comment]
By repo I mean a unity package containing the bug. Please see our creating
a bug report docs.

**Comment by Shane-SDK** [comment]
I made the simplest recreation of what is happening in my project. Unfortunately the exact location of this exception is different, however it still revolves around NetworkObjects with no GameObject being referenced. Hopefully they are rooted in the same issue.

Tested with Unity 2023.2.20f1
4.5.2R - Error occurs
4.5.0R - Error does not occur

Once imported, open **Assets/Init - OPEN ME** and start the scene. 
GUI buttons will appear
Click **Begin Host**
then click **Load Scene2**
then click **Scene1**
UnityEngine.MissingReferenceException should be raised in SceneManager.cs - Line 1640

[FishNet_Issue_803.zip](https://github.com/user-attachments/files/17563699/FishNet_Issue_803.zip)

![exception](https://github.com/user-attachments/assets/cedd86e1-babb-46d7-ba79-2f77aadbcaab)

**Comment by Zelacks** [comment]
Ive been debugging this, I hope I am not out of line but, I think I found the cause.
![image](https://github.com/user-attachments/assets/76c6fdb3-a093-4fd2-86f0-2b38dc1bbea1)
This variable guards against despawning objects that have not been initialised inside OnDestroy. It gets set on prefabs (usually) when playmode starts.
![image](https://github.com/user-attachments/assets/07eb71bb-a971-4960-a291-cb8578b13171)

BUT, instantiated prefabs do not copy over the _initalizedValusSet since it non-serialized. So destroy code is not being called on spawned instantiated prefabs

**Comment by KeithSwanger** [comment]
Can confirm that my errors also go away after removing the guard clause in `OnDestroy()`.

**Comment by FirstGearGames** [comment]
That could have other consequences but I'll get it resolved. Thanks!

**Comment by FirstGearGames** [comment]
I believe this is the actual cultprit inside `Awake` of `NetworkObject.cs`

```cs
            if (NetworkBehaviours == null || NetworkBehaviours.Count == 0)
```
The above condition is passing because values get serialized into the prefabs, this blocks running initialization as it was already done. But the initialized state is not serialized as someone said, so it never gets set true, and thus OnDestroy fails. 

Here are the changes to please try...these all take place in NetworkObject.cs

First make sure you uncomment this in OnDestroy
```cs
            //The object never initialized for use.
            if (!_initializedValusSet)
                return;
```

Change this field to Serialized/Hidden
```cs
        [SerializeField, HideInInspector]
        private bool _initializedValusSet;
```

In `Awake` change this line...
```cs
if (NetworkBehaviours == null || NetworkBehaviours.Count == 0)
```
to this...
```cs
if (!_initializedValusSet)
```

I tested this in a few projects and do not see any issues but if I could have others give it a go too and let me know that would be very helpful.

**Comment by KeithSwanger** [comment]
This seems to be working for me!

**Comment by Shane-SDK** [comment]
I updated to 4.5.2R, made the changes, and the errors have disappeared. If I experience any problems I will let everybody here know; thank you for the timely responses!

**Comment by FirstGearGames** [comment]
Thank you both for the quick test results.

**Comment by FirstGearGames** [comment]
Resolved in 4.5.4
